src/wh_server.c

```diff
@@ -28,8 +30,613 @@
+            /* check outLen */
+            if (server->cache[i].meta->len > outLen)
+                return BUFFER_E;
+            XMEMCPY((uint8_t*)meta, (uint8_t*)server->cache[i].meta,
+                sizeof(NvmMetaData));
+            XMEMCPY(out, server->cache[i].buffer, meta->len);
+            return 0;
+        }
+    }
+#if 0
+    /* setup address */
+    if (ctx->partition == 1)
+        addr = WOLFHSM_PART_1;
+    /* find id in list */
+    for (i = 0; i < WOLFHSM_KEYSLOT_COUNT; i++) {
+        /* check id */
+        if (ctx->nvmMetaCache[i].id == searchId)
+            break;
+        /* keep track of length used by previous keys */
+        keyOffset += ctx->nvmMetaCache[i].len;
+    }
+    /* if we looped to the end, no match */
+    if (i >= WOLFHSM_KEYSLOT_COUNT)
+        ret = BAD_FUNC_ARG;
+    /* check the length, meta->len is the return size */
+    if (ret == 0 && ctx->nvmMetaCache[i].len > outLen)
+        ret = BUFFER_E;
+    /* read key */
+    if (ret == 0) {
+        ret = hal_flash_read(addr + WOLFHSM_HEADER_SIZE + keyOffset, out,
+            ctx->nvmMetaCache[i].len);
+    }
+    /* calculate key digest */
+    if (ret == 0)
+        ret = wc_InitSha256_ex(sha, ctx->heap, ctx->devId);
+    if (ret == 0)
+        ret = wc_Sha256Update(sha, out, ctx->nvmMetaCache[i].len);
+    if (ret == 0)
+        ret = wc_Sha256Final(sha, digest);
+    /* validate key */
+    if (XMEMCMP(digest, ctx->nvmMetaCache[i].confDigest, WOLFHSM_DIGEST_STUB)
+        != 0) {
+        ret = BAD_FUNC_ARG;
+    }
+    if (ret == 0) {
+        /* cache key if free slot, will only kick out other commited keys */
+        hsmCacheKey(ctx, &ctx->nvmMetaCache[i], out);
+        /* update out meta with nvm cache contents */
+        XMEMCPY(meta, &ctx->nvmMetaCache[i], sizeof(NvmMetaData));
+    }
+#ifdef WOLFHSM_SHE_EXTENSION
+    /* use empty string if we couldn't find the master ecu key */
+    if (ret != 0 && searchId == WOLFHSM_SHE_MASTER_ECU_KEY_ID) {
+        XMEMSET(out, 0, WOLFHSM_SHE_KEY_SZ);
+        meta->len = WOLFHSM_SHE_KEY_SZ;
+        meta->id = searchId;
+        ret = 0;
+    }
+#endif
+#endif
+    return ret;
+}
+
+static int hsmCacheKeyCurve25519(whServerContext* server, curve25519_key* key)
+{
+    int ret;
+    uint32_t privSz = CURVE25519_KEYSIZE;
+    uint32_t pubSz = CURVE25519_KEYSIZE;
+    NvmMetaData meta[1] = {0};
+    byte keyBuf[CURVE25519_KEYSIZE * 2];
+    /* store public, then private so that loading an external public only key
+     * will work along with our keys */
+    ret = wc_curve25519_export_key_raw(key, keyBuf + CURVE25519_KEYSIZE,
+        &privSz, keyBuf, &pubSz);
+    /* cache key */
+    if (ret == 0) {
+        ret = hsmGetUniqueId(server);
+    }
+    if (ret > 0) {
+        XMEMSET(meta, 0, sizeof(NvmMetaData));
+        meta->len = privSz + pubSz;
+        meta->id = ret;
+        ret = hsmCacheKey(server, meta, keyBuf);
+    }
+    if (ret == 0)
+        ret = meta->id;
+    return ret;
```

Comment:
I think we are mixing up return codes and the unique ID. It should probably return an error code and set the ID as an output argument. We need to be able to return an error code in case the wolfcrypt calls fail etc.

Reply:
errors are negative, all keyIds are positive and are limited to uint_16 so there's no overlap